hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/resolver/MembershipNamenodeResolver.java

```diff
@@ -478,4 +478,40 @@ private List<MembershipState> getRecentRegistrationForQuery(
   public void setRouterId(String router) {
     this.routerId = router;
   }
+
+  /**
+   * Rotate cache, make the current namenode have the lowest priority,
+   * to ensure that the current namenode will not be accessed first next time.
+   *
+   * @param nsId name service id
+   * @param namenode namenode contexts
+   */
+  @Override
+  public synchronized void rotateCache(String nsId, FederationNamenodeContext namenode) {
+    cacheNS.compute(Pair.of(nsId, false), (ns, namenodeContexts) -> {
```

Comment:
I think we need to refresh the cache for the key `Pair.of(nsId, true)` at times.

You can extend the signature of rotateCache to 
```java
public synchronized void rotateCache(String nsId, FederationNamenodeContext namenode, boolean listObserversFirst)
```

And then call it with
```java
this.namenodeResolver.rotateCache(nsId, namenode, shouldUseObserver);
```